lib/util.c

```diff
@@ -312,3 +312,39 @@ doubles_almost_equal(double a, double b, double eps)
 {
     return (fabs(a) < eps && fabs(b) < eps) || gsl_fcmp(a, b, eps) == 0;
 }
+
+/* Inputs:
+ *   - `lengths` is an array of the real number lengths of `num_intervals`
+ *      consecutive half-open intervals on the real number line start at zero,
+ *      i.e. given lengths L_i, the intervals are:
+ *          [0.0, L_0), [L_0, L_0 + L_1), [L_0 + L_1, L_0 + L_1 + L_2), ...
+ *   - `x` is a real number to find in one of the intervals or after all of them
+ * Returns:
+ *   - the index (starting at zero) of the interval in which `x` was found
+ *   - `num_intervals` if `x` is at or past the end of the last interval
+ *     (the sum of all the lengths)
+ *   - zero if `x` is negative
+ *   - `num_interval` if `x` is NaN
+ * Pre-conditions:
+ *   - `lengths` must point to `num_interval` non-negative non-NaN doubles
+ * Post-condition:
+ *   - returned value is less than or equal to `num_intervals` (and non-negative)
+ *
+ */
+size_t
```

Comment:
We're not actually using this in the test suite or externally, so the usual convention would be to make it static. Then we could inline it and avoid needing to put C code in the header for the same call overhead as we have now.